src/algorithms/neal2_algorithm.cpp

```diff
@@ -19,6 +22,67 @@ Eigen::VectorXd Neal2Algorithm::lpdf_marginal_component(
   return temp_hier->marg_lpdf_grid(grid);
 }
 
+Eigen::VectorXd Neal2Algorithm::get_cluster_prior_mass(
+    const unsigned int data_idx) const {
+  unsigned int n_data = data.rows();
+  unsigned int n_clust = unique_values.size();
+  Eigen::VectorXd logprior(n_clust + 1);
+  if (mixing->is_dependent()) {
+    auto mixcast = std::dynamic_pointer_cast<DependentMixing>(mixing);
+    for (size_t j = 0; j < n_clust; j++) {
+      // Probability of being assigned to an already existing cluster
+      logprior(j) = mixcast->mass_existing_cluster(
+          unique_values[j], mix_covariates.row(data_idx), n_data - 1, true,
+          true);
+    }
+    // Further update with marginal component
+    logprior(n_clust) = mixcast->mass_new_cluster(
+        mix_covariates.row(data_idx), n_clust, n_data - 1, true, true);
+  } else {
+    for (size_t j = 0; j < n_clust; j++) {
+      // Probability of being assigned to an already existing cluster
+      logprior(j) = mixing->mass_existing_cluster(unique_values[j], n_data - 1,
+                                                  true, true);
+    }
+    // Further update with marginal component
+    logprior(n_clust) =
+        mixing->mass_new_cluster(n_clust, n_data - 1, true, true);
+  }
+  return logprior;
+}
+
+Eigen::VectorXd Neal2Algorithm::get_cluster_lpdf(
+    const unsigned int data_idx) const {
+  unsigned int n_data = data.rows();
+  unsigned int n_clust = unique_values.size();
+  Eigen::VectorXd loglpdf(n_clust + 1);
+  if (unique_values[0]->is_dependent()) {
+    auto hiercast =
```

Comment:
I would to the same as in `BaseAlgorithm::add_datum_to_hierarchy`, i.e. create a separate auxiliary method for this casting and calling the correct method (with the correct number of arguments)

For instance

```
double BaseAlgorithm::get_cluster_lpdf(int data_idx, ind cluster_idx) {
   if (unique_values[0]->is_dependent()) {
   .....
   } else {
   ....
   }
}
```

Reply:
Great idea! Will implement it ASAP

Reply:
Actually I was thinking; if we introduce changes to the hierarchies in the next PR, we can do the same for mixings, and thus doing this right now is kind of pointless

Reply:
Sure :)